### ISSUE TITLE
Non-Max Suppression

### ISSUE THREAD

**Issue by Vincent9797** [fizyr/keras-retinanet#1332]
How do I change the parameters for non-max suppression?

Currently getting an output like this:
![image](https://user-images.githubusercontent.com/42912553/79337057-67e27c80-7f57-11ea-82f6-d280c89eede0.png)
As seen above, a '5' is predicted on the leftmost character initially. On top of that, an 'S' is predicted at almost the same location. Hoping this can be removed by finetuning the setting for NMS.

My code is similar to what was provided in the Jupyter notebook:

```
from keras_retinanet.utils.image import read_image_bgr, preprocess_image, resize_image
from keras_retinanet.utils.visualization import draw_box
from keras_retinanet.utils.colors import label_color
import cv2
import time
import numpy as np
import matplotlib.pyplot as plt

# load image
image = read_image_bgr('/content/drive/My Drive/Wanted/images/test/carplatetest_frame126400_0.jpg')

# copy to draw on
draw = image.copy()
draw = cv2.cvtColor(draw, cv2.COLOR_BGR2RGB)

# preprocess image for network
image = preprocess_image(image)
image, scale = resize_image(image)

# process image
start = time.time()
boxes, scores, labels = model.predict_on_batch(np.expand_dims(image, axis=0))
print("processing time: ", time.time() - start)

# correct for image scale
boxes /= scale

labels_to_names = {}

for index, i in enumerate('0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ'):
    labels_to_names[index] = i

# visualize detections
for box, score, label in zip(boxes[0], scores[0], labels[0]):
    # scores are sorted so we can break
    if score < 0.5:
        break
        
    color = label_color(label)
    
    b = box.astype(int)
    draw_box(draw, b, color=color)
    
    print(labels_to_names[label])
    caption = "{}".format(labels_to_names[label], score)
    draw_caption(draw, b, caption)

    plt.imshow(draw)
    plt.show()
```

**Comment by Mixelon-tera** [comment]
I recommend you to apply the NMS function (not from source) at post-processing. I think the NMS function of the source has a problem.

1. Create the NMS function
def non_max_suppression_fast(boxes, overlapThresh):
    if len(boxes) == 0:
        return []
    if boxes.dtype.kind == "i":
        boxes = boxes.astype("float")
    pick = []
    x1 = boxes[:,0]
    y1 = boxes[:,1]
    x2 = boxes[:,2]
    y2 = boxes[:,3]
    area = (x2 - x1 + 1) * (y2 - y1 + 1)
    idxs = np.argsort(y2)
    while len(idxs) > 0:
        last = len(idxs) - 1
        i = idxs[last]
        pick.append(i)
        xx1 = np.maximum(x1[i], x1[idxs[:last]])
        yy1 = np.maximum(y1[i], y1[idxs[:last]])
        xx2 = np.minimum(x2[i], x2[idxs[:last]])
        yy2 = np.minimum(y2[i], y2[idxs[:last]])
       
        w = np.maximum(0, xx2 - xx1 + 1)
        h = np.maximum(0, yy2 - yy1 + 1)
       
        overlap = (w * h) / area[idxs[:last]]
        idxs = np.delete(idxs, np.concatenate(([last],
            np.where(overlap > overlapThresh)[0])))
    return boxes[pick].astype("int")

#################################################
2. Apply this function before for loop

boxes = non_max_suppression_fast(boxes[0], 0.1)
for box, score, label in zip(boxes, scores[0], labels[0]):
     ....
     ...
     ..
     .

**Comment by Vincent9797** [comment]
Thanks for your suggestion. Will close this issue.

**Comment by amardeepjaiman** [comment]
this NMS function giving inaccurate results in case of multiple classes.
